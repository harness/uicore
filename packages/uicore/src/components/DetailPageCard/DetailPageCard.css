/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.main {
  flex: 1;
  height: fit-content;
  overflow-x: hidden;

  hr {
    width: 100%;
    border-top: 1px line var(--grey-300);
  }

  .detailsSectionRowWrapper {
    padding: var(--spacing-4) 0 !important;

    &:not(:last-child) {
      border-bottom: 1px solid var(--grey-200);
    }
  }
}
