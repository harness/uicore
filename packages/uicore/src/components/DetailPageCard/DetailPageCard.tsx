/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { Container } from '../Container/Container'
import { Card } from '../Card/Card'
import { Text } from '../Text/Text'
import { Layout } from '../../layouts/Layout'
import { Color } from '@harness/design-system'
import React from 'react'
import { isUndefined } from 'lodash-es'
import cx from 'classnames'
import css from './DetailPageCard.css'
import { FontVariation } from '@harness/design-system'

export enum ContentType {
  TEXT = 'TEXT', // default
  CUSTOM = 'CUSTOM'
}

export interface Content {
  type?: ContentType
  label: string
  value?: string | JSX.Element
  newTitle?: string | JSX.Element // insert new section if there's a new title
  hideOnUndefinedValue?: boolean
}

export interface DetailPageCardProps {
  title: string | JSX.Element
  content: Content[]
  classname?: string
}

export const renderItem = ({
  type = ContentType.TEXT,
  label,
  value,
  hideOnUndefinedValue,
  newTitle,
  ...rest
}: Content): JSX.Element | undefined => {
  if (isUndefined(value) && hideOnUndefinedValue) {
    return
  }

  let jsxContent = newTitle ? (
    <Text
      style={{ marginBottom: 'var(--spacing-6)' }}
      color={Color.BLACK}
      font={{ variation: FontVariation.CARD_TITLE }}>
      {newTitle}
    </Text>
  ) : null

  if (type === ContentType.TEXT && typeof value === 'string') {
    jsxContent = (
      <>
        {jsxContent}
        <Layout.Vertical spacing="small" {...rest}>
          <Text font={{ variation: FontVariation.FORM_LABEL }} className="label">
            {label}
          </Text>
          <Text font={{ variation: FontVariation.FORM_INPUT_TEXT }} width="424px" lineClamp={1}>
            {value}
          </Text>
        </Layout.Vertical>
      </>
    )
  } else if (type === ContentType.CUSTOM) {
    jsxContent = (
      <>
        {jsxContent}
        <Layout.Vertical spacing="small" {...rest}>
          {label ? (
            <Text font={{ variation: FontVariation.FORM_LABEL }} className="label">
              {label}
            </Text>
          ) : null}
          <Text font={{ variation: FontVariation.FORM_INPUT_TEXT }} width="424px" lineClamp={1}>
            <Container className="customValue">{value}</Container>
          </Text>
        </Layout.Vertical>
      </>
    )
  }
  return <Container className={css.detailsSectionRowWrapper}>{jsxContent}</Container>
}

export const DetailPageCard: React.FC<DetailPageCardProps> = props => {
  const { title, content = [], classname } = props
  return (
    <Card className={cx(css.main, classname)} interactive={false} elevation={0} selected={false}>
      <Text color={Color.BLACK} font={{ variation: FontVariation.CARD_TITLE }}>
        {title}
      </Text>
      <Layout.Vertical style={{ marginTop: 'var(--spacing-4)' }}>
        {content.map(item => renderItem(item))}
      </Layout.Vertical>
    </Card>
  )
}
