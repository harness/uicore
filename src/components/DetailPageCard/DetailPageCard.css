.main {
  flex: 1;
  height: fit-content;
  overflow-x: hidden;

  hr {
    width: 100%;
    border-top: 1px line var(--grey-300);
  }

  .detailsSectionRowWrapper {
    padding: var(--spacing-4) 0 !important;

    &:not(:last-child) {
      border-bottom: 1px solid var(--grey-200);
    }
  }
}
