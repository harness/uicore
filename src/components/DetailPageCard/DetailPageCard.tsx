import { Container } from '../Container/Container'
import { Card } from '../Card/Card'
import { Text } from '../Text/Text'
import { Layout } from '../../layouts/Layout'
import { Color } from '../../core/Color'
import React from 'react'
import { isUndefined } from 'lodash-es'
import cx from 'classnames'
import css from './DetailPageCard.css'
import { FontVariation } from '../../styled-props/font/FontProps'

export enum ContentType {
  TEXT = 'TEXT', // default
  CUSTOM = 'CUSTOM'
}

export interface Content {
  type?: ContentType
  label: string
  value?: string | JSX.Element
  hideOnUndefinedValue?: boolean
}

export interface DetailPageCardProps {
  title: string
  content: Content[]
  classname?: string
}

export const renderItem = ({
  type = ContentType.TEXT,
  label,
  value,
  hideOnUndefinedValue,
  ...rest
}: Content): JSX.Element | undefined => {
  if (isUndefined(value) && hideOnUndefinedValue) {
    return
  }

  let jsxContent
  if (type === ContentType.TEXT) {
    jsxContent = (
      <Layout.Vertical spacing="small" {...rest}>
        <Text font="small">{label}</Text>
        <Text color={Color.BLACK} width="424px" lineClamp={1}>
          {value}
        </Text>
      </Layout.Vertical>
    )
  } else if (type === ContentType.CUSTOM) {
    jsxContent = (
      <>
        {label ? <Text font="small">{label}</Text> : null}
        {value}
      </>
    )
  }
  return <Container className={css.detailsSectionRowWrapper}>{jsxContent}</Container>
}

export const DetailPageCard: React.FC<DetailPageCardProps> = props => {
  const { title, content = [], classname } = props
  return (
    <Card className={cx(css.main, classname)} interactive={false} elevation={0} selected={false}>
      <Text color={Color.BLACK} font={{ variation: FontVariation.H5 }}>
        {title}
      </Text>
      <Layout.Vertical style={{ marginTop: 'var(--spacing-4)' }}>
        {content.map(item => renderItem(item))}
      </Layout.Vertical>
    </Card>
  )
}
