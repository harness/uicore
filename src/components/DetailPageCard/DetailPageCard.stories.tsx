import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { ContentType, DetailPageCard } from './DetailPageCard'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary } from '@storybook/addon-docs/blocks'
import { DetailPageCardProps } from './DetailPageCard'
import { Checkbox } from '../..'

export default {
  title: 'Components / DetailPageCard',
  component: DetailPageCard,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        type: 'code'
      },
      page: function PageDescription() {
        return (
          <>
            <Title>DetailPageCard</Title>
            <Subtitle>
              <pre>
                <code>{`import { DetailPageCard }  from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>
            <Primary />
            <ArgsTable story={PRIMARY_STORY} />
            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta

export const Basic: Story<DetailPageCardProps> = args => {
  return (
    <>
      <DetailPageCard {...args} />
    </>
  )
}

Basic.args = {
  title: 'Main Title',
  content: [
    { label: 'Name', value: 'Cluster Name' },
    { label: 'GitOps Agent ', value: 'Agent 1' },
    { label: 'Cache Info ', value: 'Cache 1', newTitle: 'Subtitle 1' },
    { label: 'Name', value: 'Cluster Name' },
    {
      label: 'GitOps Agent ',
      value: <Checkbox label="selected" />,
      type: ContentType.CUSTOM,
      newTitle: 'Subtitle 2'
    }
  ]
}
