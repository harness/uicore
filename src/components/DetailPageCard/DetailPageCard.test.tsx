/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { DetailPageCard } from './DetailPageCard'
describe('Detail Page card', () => {
  test('should render DetailPagecard with title and content', () => {
    const { container } = render(
      <DetailPageCard
        title="Overview"
        content={[
          { label: 'Name', value: 'Cluster Name' },
          { label: 'GitOps Agent ', value: 'Agent 1' }
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
