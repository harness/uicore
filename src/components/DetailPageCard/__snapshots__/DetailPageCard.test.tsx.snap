// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Page card should render DetailPagecard with title and content 1`] = `
<div>
  <div
    class="bp3-card bp3-elevation-0 card main"
  >
    <p
      class="font main color color-black font-variation font-variation-h5"
    >
      Overview
    </p>
    <div
      class="vertical layout-spacing-undefined main style style-marginTop style-marginTop-var(--spacing-4)"
    >
      <div
        class="font main detailsSectionRowWrapper"
      >
        <div
          class="vertical layout-spacing-small main spacing spacing-small"
        >
          <p
            class="font main font-small"
          >
            Name
          </p>
          <p
            class="font lineclamp main color color-black width width-424px"
            style="--text-line-clamp: 1; width: 424px;"
          >
            Cluster Name
          </p>
        </div>
      </div>
      <div
        class="font main detailsSectionRowWrapper"
      >
        <div
          class="vertical layout-spacing-small main spacing spacing-small"
        >
          <p
            class="font main font-small"
          >
            GitOps Agent 
          </p>
          <p
            class="font lineclamp main color color-black width width-424px"
            style="--text-line-clamp: 1; width: 424px;"
          >
            Agent 1
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
